### PR TITLE
feat!: use partshashes to create and verify proofs/roots for partset …

### PR DESCRIPTION
**Pull request by evan-forbes**
just need to fix tests before merging

### CODE REVIEW

File: consensus/propagation/types/types.go [hunk 1]
```diff
@@ -57,28 +59,42 @@ type CompactBlock struct {
...
 }
 
 // ValidateBasic checks if the CompactBlock is valid. It fails if the height is
 // negative, if the round is negative, if the BpHash is invalid, or if any of
 // the Blobs are invalid.
 func (c *CompactBlock) ValidateBasic() error {
+	err := c.Proposal.ValidateBasic()
+	if err != nil {
+		return err
+	}
```
Comment (evan-forbes) [inline consensus/propagation/types/types.go:76]: :sweat_smile: unrelated to this PR, but we forgot to check the proposal!

File: consensus/propagation/commitment.go [hunk 2]
```diff
@@ -119,15 +157,23 @@ func (blockProp *Reactor) handleCompactBlock(cb *proptypes.CompactBlock, peer p2
...
 			blockProp.Logger.Error("failed to add locally recovered part", "err", err)
 			continue
 		}
+
 		if !added {
 			blockProp.Logger.Error("failed to add locally recovered part", "part", part.Index)
 			continue
 		}
+
+		// todo: broadcast haves
```
Comment (rach-id) [inline consensus/propagation/commitment.go:176]: so we will broadcast haves for locally recovered parts? what the reasoning behind this?

File: consensus/propagation/types/types.go [hunk 3]
```diff
@@ -98,6 +114,55 @@ func (c *CompactBlock) ToProto() *protoprop.CompactBlock {
...
+
+	total := c.Proposal.BlockID.PartSetHeader.Total
+
+	c.proofsCache = make([]*merkle.Proof, 0, len(c.PartsHashes))
+
+	root, proofs := merkle.ProofsFromByteSlices(c.PartsHashes[:total])
+	c.proofsCache = append(c.proofsCache, proofs...)
+
+	if !bytes.Equal(root, c.Proposal.BlockID.PartSetHeader.Hash) {
+		return c.proofsCache, fmt.Errorf("incorect PartsHash: original root")
```
Comment (rach-id) [inline consensus/propagation/types/types.go:138]:
```suggestion
		return c.proofsCache, fmt.Errorf("incorrect PartsHash: original root")
```

File: consensus/propagation/types/types.go [hunk 4]
```diff
@@ -98,6 +114,55 @@ func (c *CompactBlock) ToProto() *protoprop.CompactBlock {
 	}
 }
 
+// Proofs returns the proofs to each part. If the proofs are not already
+// generated, then they are done so during the firstq call. An error is only
```
Comment (rach-id) [inline consensus/propagation/types/types.go:118]:
```suggestion
// generated, then they are done so during the first call. An error is only
```

**Review by rach-id — APPROVED** [review]
left non blocking feedback that can be handled in a separate PR if needed

File: consensus/propagation/commitment.go (hunk 2 above)
Comment (evan-forbes) [reply consensus/propagation/commitment.go:176]: just so that we distribute the block faster. provided we have the https://github.com/celestiaorg/celestia-core/issues/1654 working, it should be secure

File: consensus/propagation/commitment.go [hunk 5]
```diff
@@ -69,15 +69,44 @@ func (blockProp *Reactor) ProposeBlock(proposal *types.Proposal, block *types.Pa
 	}
 }
 
+func extractHashes(blocks ...*types.PartSet) [][]byte {
+	total := uint32(0)
+	for _, block := range blocks {
+		total += block.Total()
+	}
+
+	partHashes := make([][]byte, 0, total) // Preallocate capacity
```
[the remark this answers is not in the document]
Comment (evan-forbes) [reply consensus/propagation/commitment.go:78]: why is that? 

that comment can be removed however (damn llm)

File: consensus/propagation/types/types.go (hunk 3 above)
Comment (evan-forbes) [reply consensus/propagation/types/types.go]: good eye, updated elsewhere

File: consensus/propagation/commitment_test.go [hunk 6]
```diff
@@ -131,57 +131,6 @@ func TestRecoverPartsLocally(t *testing.T) {
...
-	// skew the part bytes
-	partSet.GetPart(2).Bytes[10] = 0x12
-
-	blockPropR.ProposeBlock(prop, partSet, metaData)
-
-	_, actualParts, _ := blockPropR.GetProposal(prop.Height, prop.Round)
-
-	// verify that the part 2 was not added to the part set
-	assert.Nil(t, actualParts.GetPart(2))
-}
```
Comment (evan-forbes) [inline consensus/propagation/commitment_test.go:181]: @rach-id I think we can delete this, as I think the only thing that it was testing is that the parts the proposer match the hashes, which isn't actually a property we need to test since we're not handling that error. Instead, we should be testing that recovery parts that don't match the hash are not propagated

Comment (rach-id) [reply consensus/propagation/commitment_test.go:181]: it doesn't return an error, but it's not added to the partset. This test check whether an invalid part was added to the partset. This can happen in some edge case where the tx positions or the unmarshalling of the bytes are invalid for some reason, this test makes sure we're not adding any invalid data to the partset.
